### PR TITLE
[infra-proxy-service] Add validation for servers & orgs

### PR DESCRIPTION
**Pull request by vsingh-msys**
### :nut_and_bolt: Description: What code changed, and why?

This PR is the initial draft for validation services. Wrote some of the validation for the servers API requests.

For Advance, GRPC also get integrated but required major changes at UI end. so 

added `NewPlainError` func to raise sequencing error.

### :chains: Related Resources
Fixes https://github.com/chef/automate/issues/3296

### :+1: Definition of Done
 - pre validate request with required validation rules.

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
Signed-off-by: Vivek Singh <vivek.singh@msystechnologies.com>

### CODE REVIEW

File: components/infra-proxy-service/validation/validation.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	return &Validator{Opts: opts}
+}
+
+// Validate msg
+func (v *Validator) Validate() error {
+	reqVal := reflect.ValueOf(v.Opts.Request)
+	typeOfS := reqVal.Type()
+	if v.Opts.RequiredDefault {
+		for i := 0; i < reqVal.NumField(); i++ {
+			err1 := RequiredField(reqVal.Field(i).String(), strings.Split(typeOfS.Field(i).Tag.Get("json"), ",")[0], v.Opts.Target)
```
Comment (srenatus) [inline components/infra-proxy-service/validation/validation.go:50]: [nit] It's common to just always use `err`. Is there a reason not to, here?

Comment (vsingh-msys) [reply components/infra-proxy-service/validation/validation.go:50]: Nope, there is no specific reason, just left while doing the cleanup. Thanks

**Comment by vsingh-msys** [comment]
Thanks for pointing out about `FieldViolation` and bad request error. I have seen some of the directions in one valuation PR surely will into it.


As this PR not yet finalize and just working on various aspects of providing rules for validations hope will come with the final solution soon.

**Comment by bcmdarroch** [comment]
I've been working on adding a shared validation lib: https://github.com/chef/automate/pull/3502

Would be interested in seeing this code in the library, especially if we can come up with a more general solution that works across all our services. Maybe we could meet sometime next week to discuss?

**Comment by vsingh-msys** [comment]
Added google rpc error_details proto, not it's working fine, with the details error message, still need some of the decision to make!

New error format:

```
curl -sSkH "api-token: $(get_admin_token)" https://a2-dev.test/api/v0/infra/serverspretty -X POST -d '{}'
{
  "error": "invalid argument",
  "code": 3,
  "message": "invalid argument",
  "details": [
    {
      "@type": "type.googleapis.com/google.rpc.BadRequest",
      "field_violations": [
        {
          "field": "name",
          "description": "a server name is required and must contain at least one non-whitespace character"
        },
        {
          "field": "description",
          "description": "a server description is required and must contain at least one non-whitespace character"
        },
        {
          "field": "fqdn",
          "description": "a server fqdn is required and must contain at least one non-whitespace character"
        },
        {
          "field": "ip_address",
          "description": "a server ip_address is required and must contain at least one non-whitespace character"
        }
      ]
    }
  ]
}

```

**Comment by vjeffrey** [comment]
@vsingh-msys should this be closed?

**Comment by vsingh-msys** [comment]
Not now, still pending from my side, will takeup soon thanks!

File: components/infra-proxy-service/validation/validation.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		}
+	}
+
+	for field, rules := range v.Opts.Rules {
+		for _, rule := range rules {
+			switch rule {
+			case "required":
+				fv := RequiredField(getFieldValue(reqVal, field), getFieldTag(typeOfS, field), v.Opts.Target)
+				if fv != nil {
+					return NewPlainError(fv)
```
Comment (vsingh-msys) [inline components/infra-proxy-service/validation/validation.go:57]: Raise only sequential errors to keep parity with existing UI error parsing.

**Pull request merged**